gpkit/tools/docstring.py

```diff
@@ -57,17 +58,62 @@ def expected_unbounded(instance, doc):
     return exp_unbounded
 
 
-def parse_variables(string, errorcatch=True):
+class parse_variables(object):  # pylint:disable=invalid-name
+    """decorator for adding local Variables from a string.
+
+    Generally called as `@parse_variables(__doc__, globals())`.
+    """
+    def __init__(self, string, scopevars=None):
+        self.string = string
+        self.scopevars = scopevars
+        if scopevars is None:
+            raise DeprecationWarning("""
+parse_variables is no longer used directly with exec, but as a decorator:
+
+    @parse_variables(__doc__, globals())
+    def setup(...):
+
+""")
+
+    def __call__(self, function):  # pylint:disable=too-many-locals
+        orig_lines, lineno = inspect.getsourcelines(function)
+        indent_length = 0
+        while orig_lines[1][indent_length] in [" ", "\t"]:
+            indent_length += 1
+        first_indent_length = indent_length
+        next_indented_idx = 2
+        # get the next indented line
+        while len(orig_lines[next_indented_idx]) <= indent_length + 1:
+            next_indented_idx += 1
+        while orig_lines[next_indented_idx][indent_length] in [" ", "\t"]:
+            indent_length += 1
+        second_indent = orig_lines[next_indented_idx][:indent_length]
+        parse_lines = [second_indent + line + "\n"
+                       for line in parse_varstring(self.string).split("\n")]
+        # make ast of these new lines, insert it into the original ast
+        new_lines = [orig_lines[1]] + parse_lines + orig_lines[2:]
+        new_src = "\n".join([l[first_indent_length:-1] for l in new_lines])
+        new_ast = ast.parse(new_src, "<parse_variables>")
+        ast.increment_lineno(new_ast, n=lineno-len(parse_lines))
+        code = compile(new_ast, inspect.getsourcefile(function), "exec",
+                       dont_inherit=True)  # don't inherit __future__ from here
+        out = {}
+        exec(code, self.scopevars, out)  # pylint: disable=exec-used
+        return out[function.__name__]
+
+
+def parse_varstring(string):
     "Parses a string to determine what variables to create from it"
-    if sys.version_info >= (3, 0):
-        raise FutureWarning("parse_variables is not yet supported in Python 3")
-    out = "from gpkit import Variable, VectorVariable\n"
-    out += check_and_parse_flag(string, "Constants\n", errorcatch,
-                                constant_declare)
-    out += check_and_parse_flag(string, "Variables\n", errorcatch)
-    out += check_and_parse_flag(string, "Variables of length", errorcatch,
-                                vv_declare)
-    return out
+    consts = check_and_parse_flag(string, "Constants\n", constant_declare)
+    variables = check_and_parse_flag(string, "Variables\n")
+    vecvars = check_and_parse_flag(string, "Variables of length", vv_declare)
+    out = ["# " + line for line in string.split("\n")]
+    out[0] = "from gpkit import Variable, VectorVariable" + "  " + out[0]
```

Comment:
inline comment for this? Will the two imports ever change / need to be expanded?

Reply:
not unless the sections just above are changed; added a comment!